services/admin/src/store.ts

```diff
@@ -277,16 +285,55 @@ export class Store {
 
         const lines = readline.createInterface(fs.createReadStream(filePath));
 
+        const currentCvrFileMode =
+          this.getCurrentCvrFileModeForElection(electionId);
+        let lastLineCvrMode: Optional<Admin.CvrFileMode>;
+        let nextLineCvrMode: Optional<Admin.CvrFileMode>;
+
         for await (const line of lines) {
           if (line.trim() === '') {
             continue;
           }
 
-          // TODO(https://github.com/votingworks/vxsuite/issues/2716): add error checking against
-          // attempts to mix "live" and "test" CVRs for the same election.
+          // Parse CVR data for validation.
+          const cvr = safeParseJson(line).unsafeUnwrap() as CastVoteRecord;
+          nextLineCvrMode = cvr._testBallot
+            ? Admin.CvrFileMode.Test
+            : Admin.CvrFileMode.Official;
+
+          // Ensure CVR mode matches previous file imports, if any:
+          if (
+            currentCvrFileMode !== Admin.CvrFileMode.Unlocked &&
+            nextLineCvrMode !== currentCvrFileMode
+          ) {
+            this.client.run('rollback transaction');
+
+            return err({
+              kind: 'MixedLiveAndTestBallots',
+              userFriendlyMessage:
+                `this file contains ${nextLineCvrMode} ballots, ` +
+                `but you are currently in ${currentCvrFileMode} mode`,
+            });
+          }
+
+          // Ensure CVR mode matches last-processed CVR:
+          if (lastLineCvrMode && nextLineCvrMode !== lastLineCvrMode) {
+            this.client.run('rollback transaction');
+
+            return err({
+              kind: 'MixedLiveAndTestBallots',
+              userFriendlyMessage:
+                'these CVRs cannot be tabulated together because ' +
+                'they mix live and test ballots',
+            });
+          }
+
+          lastLineCvrMode = nextLineCvrMode;
+
           const result = this.addCastVoteRecordFileEntry(electionId, id, line);
 
           if (result.isErr()) {
+            this.client.run('rollback transaction');
```

Comment:
Nice catch! I wonder if there's a way to ensure we always rollback the transaction on error (other than just being diligent)

Reply:
that's worth looking into. long shot, but hoping there's a way to set sqlite transactions to time out from inactivity and roll back - at the very least, we can re-structure the code a bit

Reply:
Yeah, maybe we can have some sort of `withTransaction(doTransaction: (client) => void)` function that knows to rollback on `err` results and exceptions...

Reply:
really like this - gonna follow up with another PR to implement!